File: 1658. Minimum Operations to Reduce X to Zero.java
// Medium
// Array, Sliding Window
// O(n)

class Solution {
  public int minOperations(int[] nums, int x) {
      // calculate the total sum of nums
      int numsSum = 0;
      for(int num : nums){
          numsSum += num;
      }
      // the target sum of what we find(the max subarray) is numsSum - x
      int target = numsSum - x;
      // handle edge case
      if(target < 0) return -1;
      if (target == 0)
        return nums.length;
      
      // find the max sub array which sum is target 
      int maxLen = -1;
      int sum = 0;
      int left = 0;

      for(int right = 0; right < nums.length; right++){
          sum += nums[right];
          while (sum > target && left < right){
              sum -= nums[left];
              left++;
          }
          if(sum == target){
              maxLen = Math.max(maxLen, right - left + 1);
          }
      }
      if(maxLen == -1) return -1;
      return nums.length - maxLen;

  }
}

File: 1001-2000/1658. Minimum Operations to Reduce X to Zero.py
# Time:O(n), Space:O(1)

from typing import List

class Solution:
    def minOperations(self, nums: List[int], x: int) -> int:
        # 相当于找最长连续子数组，和为total - x
        total = sum(nums)
        target = total - x
        n = len(nums)

        if target < 0:
            return -1
        if target == 0:
            return n
        
        left = 0
        cur_sum = 0
        max_len = -1

        for right in range(n):
            cur_sum += nums[right]

            while cur_sum > target:
                cur_sum -= nums[left]
                left += 1
            
            if cur_sum == target:
                max_len = max(max_len, right - left + 1)
            
        return n - max_len if max_len != -1 else -1
    